xlators/storage/posix/src/posix-io-uring.c

```diff
@@ -0,0 +1,634 @@
+               "readv(async) failed fd=%d.", _fd);
+        goto out;
+    }
+
+    ret = posix_fdstat(this, fd->inode, _fd, &postbuf);
+    if (ret != 0) {
+        op_ret = -1;
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%d", _fd);
+        goto out;
+    }
+
+    op_ret = res;
+    op_errno = 0;
+
+    iobref = iobref_new();
+    if (!iobref) {
+        op_ret = -1;
+        op_errno = ENOMEM;
+        goto out;
+    }
+
+    iobref_add(iobref, iobuf);
+
+    iov.iov_base = iobuf_ptr(iobuf);
+    iov.iov_len = op_ret;
+
+    /* Hack to notify higher layers of EOF. */
+    if (!postbuf.ia_size || (offset + iov.iov_len) >= postbuf.ia_size)
+        op_errno = ENOENT;
+
+    GF_ATOMIC_ADD(priv->read_value, op_ret);
+    // response xdata is used only for cloudsync, so ignore for now.
+out:
+    STACK_UNWIND_STRICT(readv, frame, op_ret, op_errno, &iov, 1, &postbuf,
+                        iobref, NULL);
+    if (iobref)
+        iobref_unref(iobref);
+    posix_io_uring_ctx_free(ctx);
+}
+
+static void
+posix_prep_readv(struct io_uring_sqe *sqe, struct posix_uring_ctx *ctx)
+{
+    sqe->flags |= IOSQE_ASYNC;
+    io_uring_prep_readv(sqe, ctx->_fd, &ctx->fop.read.iovec, 1,
+                        ctx->fop.read.offset);
+}
+
+int
+posix_io_uring_readv(call_frame_t *frame, xlator_t *this, fd_t *fd, size_t size,
+                     off_t offset, uint32_t flags, dict_t *xdata)
+{
+    struct posix_uring_ctx *ctx = NULL;
+    int32_t op_errno = ENOMEM;
+    struct iobuf *iobuf = NULL;
+    int ret = 0;
+
+    ctx = posix_io_uring_ctx_init(
+        frame, this, fd, GF_FOP_READ, posix_prep_readv,
+        posix_io_uring_readv_complete, &op_errno, xdata);
+    if (!ctx) {
+        goto err;
+    }
+
+    iobuf = iobuf_get2(this->ctx->iobuf_pool, size);
+    if (!iobuf) {
+        op_errno = ENOMEM;
+        goto err;
+    }
+    ctx->fop.read.iobuf = iobuf;
+    ctx->fop.read.iovec.iov_base = iobuf_ptr(iobuf);
+    ctx->fop.read.iovec.iov_len = size;
+    ctx->fop.read.offset = offset;
+
+    ret = posix_io_uring_submit(this, ctx);
+    if (ret < 0) {
+        gf_msg(this->name, GF_LOG_ERROR, -ret, P_MSG_POSIX_IO_URING,
+               "Failed to submit sqe");
+        op_errno = -ret;
+        goto err;
+    }
+    if (ret == 0) {
+        gf_msg(this->name, GF_LOG_WARNING, -ret, P_MSG_POSIX_IO_URING,
+               "submit sqe got zero");
+    }
+    return 0;
+err:
+    STACK_UNWIND_STRICT(readv, frame, -1, op_errno, NULL, 1, NULL, NULL, NULL);
+    posix_io_uring_ctx_free(ctx);
+    return 0;
+}
+
+static void
+posix_writev_fill_rsp_dict(struct posix_uring_ctx *ctx, xlator_t *this,
+                           dict_t **rsp_xdata)
+{
+    int is_append = 0;
+
+    // For posix_inode_ctx_t->write_atomic_lock like behaviour, looks like we
+    // need to chain prestat, write and poststat into a chained sqe.
```

Comment:
A chained SQE doesn't guarantee atomic execution, only sequential execution.

Reply:
It is not clear to me if io-uring can interrupt a chained SQE in the middle to run another SQE and then come back to the first one. But I will remove this comment for now.

Reply:
Found this in the doc: 

> Chains are independent execution units, and multiple chains can execute and complete in parallel to each other. This includes sqes that are not part of any chain.

 So it does look like another  sqe (chained/ normal) on the same file or fd can run in parallel.

Reply:
Yes, chained requests are to ensure that a set of requests are processed sequentially, but there's no other guarantee. io_uring will try to execute as many requests as possible, so other SQE's may be processed in the middle.